Source/Engine/Physics/PhysicsScene.cpp

```diff
@@ -0,0 +1,818 @@
+    if (PxFilterObjectIsKinematic(attributes0) && PxFilterObjectIsKinematic(attributes1))
+    {
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_FOUND;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_PERSISTS;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_LOST;
+        pairFlags |= PxPairFlag::eDETECT_DISCRETE_CONTACT;
+        return PxFilterFlag::eSUPPRESS;
+    }
+
+    // Trigger the contact callback for pairs (A,B) where the filtermask of A contains the ID of B and vice versa
+    if ((filterData0.word0 & filterData1.word1) && (filterData1.word0 & filterData0.word1))
+    {
+        pairFlags |= PxPairFlag::eSOLVE_CONTACT;
+        pairFlags |= PxPairFlag::eDETECT_DISCRETE_CONTACT;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_FOUND;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_PERSISTS;
+        pairFlags |= PxPairFlag::ePOST_SOLVER_VELOCITY;
+        pairFlags |= PxPairFlag::eNOTIFY_CONTACT_POINTS;
+        return PxFilterFlag::eDEFAULT;
+    }
+
+    // Ignore pair (no collisions nor events)
+    return PxFilterFlag::eKILL;
+}
+
+enum class ActionType
+{
+    Sleep,
+};
+
+struct ActionData
+{
+    ActionType Type;
+    PxActor* Actor;
+};
+
+#if WITH_VEHICLE
+
+static PxQueryHitType::Enum WheelRaycastPreFilter(PxFilterData filterData0, PxFilterData filterData1, const void* constantBlock, PxU32 constantBlockSize, PxHitFlags& queryFlags)
+{
+    // Hardcoded id for vehicle shapes masking
+    if (filterData0.word3 == filterData1.word3)
+    {
+        return PxQueryHitType::eNONE;
+    }
+
+    // Collide for pairs (A,B) where the filtermask of A contains the ID of B and vice versa
+    if ((filterData0.word0 & filterData1.word1) && (filterData1.word0 & filterData0.word1))
+    {
+        return PxQueryHitType::eBLOCK;
+    }
+
+    return PxQueryHitType::eNONE;
+}
+
+#endif
+
+class PhysicsScenePhysX
+{
+    friend PhysicsScene;
+
+private:
+    PxScene* Scene;
+    PxCpuDispatcher* CpuDispatcher;
+    PxControllerManager* ControllerManager;
+    PxSimulationFilterShader PhysXDefaultFilterShader = PxDefaultSimulationFilterShader;
+    SimulationEventCallback EventsCallback;
+    CriticalSection FlushLocker;
+
+    Array<PxActor*> NewActors;
+    Array<PxActor*> DeadActors;
+    Array<PxMaterial*> DeadMaterials;
+    Array<PhysicsColliderActor*> DeadColliders;
+    Array<Joint*> DeadJoints;
+    Array<ActionData> Actions;
+    Array<PxBase*> DeadObjects;
+
+#if WITH_VEHICLE
+    Array<PxVehicleWheels*> WheelVehiclesCache;
+    Array<PxRaycastQueryResult> WheelQueryResults;
+    Array<PxRaycastHit> WheelHitResults;
+    Array<PxWheelQueryResult> WheelVehiclesResultsPerWheel;
+    Array<PxVehicleWheelQueryResult> WheelVehiclesResultsPerVehicle;
+    PxBatchQuery* WheelRaycastBatchQuery = nullptr;
+    PxVehicleDrivableSurfaceToTireFrictionPairs* WheelTireFrictions = nullptr;
+    Array<WheeledVehicle*> WheelVehicles;
+#endif
+};
+
+PhysicsScene::PhysicsScene(const String& name, const PhysicsSettings& settings)
```

Comment:
`PhysicsScene` could contain `Init` method that would take those settings and return `true` on initialization fail.

Reply:
It makes sense but I think it's out of scope for this PR but I understand the appeal of making these changes all at once.

I can do this but I'd suggest merging this first and implementing the abstraction in another PR. There are more things that need to be abstracted such as the CharacterController and Vehicle as they have direct dependencies on PhysX. I'm sure this will involve discussion and it makes sense to me to keep the issues separate.

It's up to you though, we can implement both features in this branch if you feel strongly that way.